runner/app/pipelines/segment_anything_2.py

```diff
@@ -15,22 +21,73 @@ def __init__(self, model_id: str):
+                torch.backends.cuda.matmul.allow_tf32 = True
+                torch.backends.cudnn.allow_tf32 = True
+        elif torch_device.type == "mps":
+            print(
+                "\nSupport for MPS devices is preliminary. SAM 2 is trained with CUDA and might "
+                "give numerically different outputs and sometimes degraded performance on MPS. "
+                "See e.g. https://github.com/pytorch/pytorch/issues/84936 for a discussion."
+            )
 
         self.tm = SAM2ImagePredictor.from_pretrained(
             model_id=model_id,
             device=torch_device,
         )
+        
+        self.tm_vid = SAM2VideoPredictor.from_pretrained(
+            model_id=model_id,
+            device=torch_device,
+        )
 
     def __call__(
-        self, image: PIL.Image, **kwargs
-    ) -> Tuple[List[PIL.Image], List[Optional[bool]]]:
-
-        self.tm.set_image(image)
+        self, media_file: UploadFile, media_type: str, **kwargs
+    ) -> Tuple[List[UploadFile], str, List[Optional[bool]]]:
+        if media_type == "image":
+            try:
+                image = Image.open(media_file.file)
+                self.tm.set_image(image)
+                prediction = self.tm.predict(**kwargs)
+            except Exception as e:
+                raise InferenceError(original_exception=e)
+        elif media_type == "video":
+            try:
+                temp_dir = tempfile.mkdtemp()
+                # TODO: Fix the file type dependency, try passing to ffmpeg without saving to file
+                video_path = f"{temp_dir}/input.mp4"
+                with open(video_path, "wb") as video_file:
+                    video_file.write(media_file.file.read())
+                
+                # Run ffmpeg command to extract frames from video
+                frame_dir = tempfile.mkdtemp()
+                output_pattern = f"{frame_dir}/%05d.jpg"
+                ffmpeg_command = f"ffmpeg -i {video_path} -q:v 2 -start_number 0 {output_pattern}"
+                subprocess.run(ffmpeg_command, shell=True, check=True)
+                shutil.rmtree(temp_dir) 
+                
+                inference_state = self.tm_vid.init_state(video_path=frame_dir)
+                shutil.rmtree(frame_dir)
 
-        try:
-            prediction = self.tm.predict(**kwargs)
-        except Exception as e:
-            raise InferenceError(original_exception=e)
+                _, out_obj_ids, out_mask_logits = self.tm_vid.add_new_points_or_box(
+                    inference_state,
+                    frame_idx=5,
+                    obj_id=1,
+                    points=kwargs.get('point_coords', None),
+                    labels=kwargs.get('point_labels', None),
+                    )
+                video_segments = {}  # video_segments contains the per-frame segmentation results
+                for out_frame_idx, out_obj_ids, out_mask_logits in self.tm_vid.propagate_in_video(inference_state):
```

Comment:
I have a feeling we should return the full triple instead of creating a video segment, leaving post-processing to the consumer of the API ( though I recognize this is good quick way to validate the sanity of the mask outputs )

Reply:
Thanks for the guidance on this, I see how we can just return the results of `self.tm_vid.propagate_in_video(inference_state):`

Reply:
> I have a feeling we should return the full triple instead of creating a video segment, leaving post-processing to the consumer of the API ( though I recognize this is good quick way to validate the sanity of the mask outputs )

I had some issues trying to return the correct values. I added frame index as an input parameter, normally `propagate_in_video` will loop returning results for each frame starting at the frame index until the end of the video, now it should only return a single frame. But the data doesn't look correct, can you take a look?